Add RSpec 4 compatibility

pirj opened the pull request:
You may check a PR that also switches to use RSpec 4 and RSpec-Rails 6: https://github.com/pirj/discourse/pull/1

There are a few reasons it doesn't make sense to [switch right away](https://github.com/pirj/discourse/pull/1/commits/08f058a7ee448b3286d576d4320ac1c7f365103c):
 - RSpec 4 and RSpec-Rails 6 are not yet released, and it's a hassle to [source them from GitHub](https://github.com/pirj/discourse/pull/1/files#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR160-R164)
 - RSpec-Rails 6 is [not marked to support RSpec 4](https://github.com/pirj/rspec-rails/commit/6023a2003d5b62b8619f83ae470ef31a6afcba60)
 - ~`rspec-html-matchers` won't work until [this PR](https://github.com/kucaahbe/rspec-html-matchers/pull/77) is merged~ - [fixed in 0.10.0](https://github.com/kucaahbe/rspec-html-matchers/commit/54e99fba653d4b712fd83c87c1bffbe68726f51b)
 - 61 plugins need to be updated ([example PR](https://github.com/discourse/discourse-teambuild/pull/42))

### What's Done

#### Comply to strict predicate matchers

   See:
   - https://github.com/rspec/rspec-expectations/pull/1195
   - https://github.com/rspec/rspec-expectations/pull/1196
   - https://github.com/rspec/rspec-expectations/pull/1277

Basically,
```ruby
expect(OpenStruct.new(foo?: nil)).to_not be_foo
```
won't pass anymore, as strict predicate matchers would only accept `true`/`false`, but not "truthy" or "falsey" values.

#### Use the non-globally exposed RSpec syntax

See https://github.com/rspec/rspec-core/pull/2803

`describe`/`shared_examples` won't work without the explicit `RSpec.`.
`should`/`should_not` were deprecate nearly a decade ago and will be removed.

#### Misc

Discourse specs use defaults, and it helped a lot, there's just `--color` option that had to be removed from `.rspec` file.
https://github.com/rspec/rspec-core/commit/04078317ba6577699d06cf4dccf014254dcde7a6

#### RSpec-Rails 6

No changes turned out to be necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/discourse/discourse?pullRequest=17652) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/discourse/discourse?pullRequest=17652) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/discourse/discourse?pullRequest=17652) it.</sub>

pirj commented on spec/lib/svg_sprite/svg_sprite_spec.rb:
```diff
@@ -254,7 +254,7 @@
       theme.save!
 
-      expect(Upload.where(id: upload.id)).to be_exist
+      expect(Upload.where(id: upload.id)).to be_exists
```
Because `Upload.where(...).exist?` method doesn't exist, but `exists?` does.

tgxworld replied:
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

tgxworld commented on spec/lib/svg_sprite/svg_sprite_spec.rb:
```diff
@@ -266,7 +266,7 @@
       theme.save!
 
-      expect(Upload.where(id: upload.id)).to be_exist
+      expect(Upload.where(id: upload.id)).to be_exists
```
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

tgxworld commented on spec/lib/svg_sprite/svg_sprite_spec.rb:
```diff
@@ -280,7 +280,7 @@
       child_theme.save!
 
-      expect(Upload.where(id: upload.id)).to be_exist
+      expect(Upload.where(id: upload.id)).to be_exists
```
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

tgxworld commented on spec/models/topic_spec.rb:
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
   describe '#secure_category?' do
     let(:category) { Category.new }
 
+    def read_restricted_category?(topic)
+      topic.read_restricted_category?
+    end
```
```suggestion
```

tgxworld commented on spec/models/topic_spec.rb:
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
     it "is false if the category is not secure" do
       category.stubs(:read_restricted).returns(false)
-      expect(Topic.new(category: category)).not_to be_read_restricted_category
+      expect(read_restricted_category?(Topic.new(category: category))).to be(false)
```
```suggestion
      expect(Topic.new(category: category).read_restricted_category?).to eq(false)
```

tgxworld commented on spec/models/topic_spec.rb:
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
-    it "is false if there is no category" do
-      expect(Topic.new(category: nil)).not_to be_read_restricted_category
+    it "is falsey if there is no category" do
+      expect(read_restricted_category?(Topic.new(category: nil))).to be(nil)
```
```suggestion
      expect(Topic.new(category: nil).read_restricted_category?).to eq(nil)
```

tgxworld commented:
Thank you @pirj, some comments but otherwise the PR looks good :+1:

pirj commented on spec/models/topic_spec.rb:
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
     it "is true if the category is secure" do
       category.stubs(:read_restricted).returns(true)
-      expect(Topic.new(category: category)).to be_read_restricted_category
+      expect(read_restricted_category?(Topic.new(category: category))).to be(true)
```
I'd stick with `be(true)`, but `eq(true)` is fine, too.

pirj commented:
Thanks for the quick review, @tgxworld . Do you want me to squash the last commit?

pirj commented:
Rebased to resolve conflicts, squashed code review fixups.
I dared updating `rspec-html-matchers` to a version that is compatible with RSpec 4 in the last commit, see [their update](https://github.com/kucaahbe/rspec-html-matchers/commit/54e99fba653d4b712fd83c87c1bffbe68726f51b) for details.

tgxworld commented:
Don't worry about squashing because we'll do a squash and merge once the PR is ready.

pirj commented:
> the PR is ready

Is there something I can do to help making it ready?

tgxworld commented:
@pirj Thank you for your contribution here :+1:
